### ISSUE TITLE
Proposal: Array and other classes with indexer should have batch assignment syntax for their elements

### ISSUE THREAD

**Issue by ygc369** [dotnet/csharplang#232]
Batch assignment need not alloc new memory.
```C#
//Example1:
int[] a=new int[]{2,3,5,7,11};
...... //do something
a[]={13,17,19,23,29}; //a[0]=13,a[1]=17, ...... , a[4]=29
...... //do something
a[1..3]={23,15,4}; //a[1]=23,a[2]=15,a[3]=4

//Example2:
unsafe static void func()
{
    int* prime=stackalloc int[5];
    prime[]={2,3,5,7,11}; //prime[0]=2,prime[1]=3, ...... , prime[4]=11
    ......
}

//Example3:
class A
{
    private int[] a;
    public A(int length)
    {
        this.a=new int[length];
    }
    public int this[int i] //indexer
    {
        get{
            if(a==null || i<0 || i>=a.Length)return 0;
            return a[i];
        }
        set{
             if(a==null || i<0 || i>=a.Length)return;
             a[i]=value;
        }
    }
}
.......
A t=new A(10);
t[]={3,7,10,41}; //t[0]=3, t[1]=7, t[2]=10, t[3]=41
......
t[5..]={28,6}; //t[5]=28, t[6]=6
......
t[..9]={0,33,20}; //t[7]=0,t[8]=33,t[9]=20
.......
t[0..9]={1}; //t[0]=1,t[1]=0, ...... , t[9]=0
.......
A tt; tt[]={1,2,3}; //compile error! tt has not allocated memory!
```

**Comment by ygc369** [comment]
This proposal would make code less. I think it's not hard to implement.

**Comment by chrisaut** [comment]
@ygc369  Can you please format your proposal, it's really hard to read.

**Comment by jnm2** [comment]
@ygc369 See https://guides.github.com/features/mastering-markdown/

**Comment by ygc369** [comment]
@chrisaut 
@jnm2 
OK, I've done.

**Comment by apskim** [comment]
Looks like this proposal could be merged into #198 and #185 as an additional discussion question. Your proposal seems to be that\*:

* A slice expression `a[l..r]` is allowed as an lvalue assuming the slice receiver `a` and its individual index access expressions `a[i]` are valid and writable.
  
* The compiler implicitly rewrites a write into a slice expression as an index-by-index assignment of individual elements of the slice. If the rvalue being written is a compile-time constant, it is not allocated in memory as a whole.

Some immediate questions:

1. What happens if the slice sizes don't align? I'd expect a silent success if the rvalue is shorter, a compile-time error if the rvalue is longer and known at compile time, and a run-time error if the rvalue is longer and unknown at compile time. However, there are other options.

``` cs
/* Note: I'm using inclusive-on-both-ends range semantics */
var a = new string[5];  // {"", "", "", "", ""}
a[1..4] = { "foo" };  // {"", "foo", "", "", ""}
a[1..2] = { "foo", "bar", "baz" }  // compile-time error
a[1..2] = Console.ReadLine().Split();  // run-time error if there're more than two words
```

2. Your proposal only mentions assignments. I generalized it to arbitrary lvalues for consistency, but it doesn't have to be the case. If it does, what's the exact semantics of all operations? For instance, is this allowed?

``` cs
void M(out int[] p);
void M(out Span<int> p);
void M(out IEnumerable<int> p);

var a = new int[5];
M(out a[1:3]);
// Does this compile?
// If so, which overload is called?
// In addition, which overload _can_ be called (assuming the absence of others)?
```

\* Pardon my terminology, I'm not fluent in the C# spec.

**Comment by ygc369** [comment]
@apskim 
Thank you for comment. 
Your questions seem not big problems.
Q1: Your suggestion is OK.
Q2: I suggest to disallow this kind of usage for now, before we find a best solution.
